airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/demo/file_based/iface/StreamDestinationFactory.java

```diff
@@ -0,0 +1,9 @@
+package io.airbyte.integrations.destination.snowflake.demo.file_based.iface;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import io.airbyte.integrations.base.destination.typing_deduping.StreamConfig;
+
+public interface StreamDestinationFactory extends AutoCloseable {
+  void setup(JsonNode config);
```

Comment:
There are two `setup` methods, which is slightly confusing. To recap:
* the StreamDestinationFactory setup method is to create anything that's shared by all streams. E.g. a connection pool, the `airbyte_internal` database schema, etc.
* The StreamDestination setup method is to set up a specific stream - creating its raw/final tables, etc.

Similarly, their `close` methods have the same name, but do different things.

Reply:
Can I argue for more explicit life-cycle management (complete with logging)?
I'd go with:
* Initialize
  * Boot up
  * Parse Config
  * Get logger going
  * Don't do anything over the network
* Start 
  * Main process start
  * Connect to DB / Make connection Pool
  * Do CHECK connection stuf
  * Do shared setup (airbyte_internal schema)
* Run
  * Make multiple threads
  * Do the sync
* Shutdown
  * Final T&D
  * Checksum / Stats about the sync
* Stop
  * Disconnect from network things / close DB pool
  * exit(0)

Reply:
seems reasonable to me. The exact contents of those methods might be different (e.g. file destinations would skip / tweak some of the subbullets) but I like the explicit separation between e.g. pure config/catalog parsing vs actually doing stuff in the destination

I also realized that (since we do need to check for conflicting names) that the global init phase needs some way to detect those conflicts + rename tables as needed. Not going to add that to this PR, but probably something like StreamDestination returning a generic `T`, and being able to accept a stream name + suffix